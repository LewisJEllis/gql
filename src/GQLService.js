/* @flow */
import * as schema from './schema';
import * as query from './query';

import GQLConfig from './config/GQLConfig';
import GQLWatcher from './shared/GQLWatcher';
import debug from './shared/debug';

import type { GQLHint, GQLInfo, DefLocation, Position } from './shared/types';

import type { GQLError } from './shared/GQLError';

type Options = $Exact<{
  cwd?: string,
  onChange?: () => void,
  onInit?: () => void,
  onError?: () => void,
  debug?: boolean,
  watchman?: boolean,
  watch?: boolean,
}>;

type CommandParams = $Exact<{
  sourcePath: string,
  sourceText: string,
  position: Position,
}>;

export type { Options as GQLServiceOptions };

const _noop = () => {}; // eslint-disable-line no-empty-function

export class GQLService {
  _isInitialized: boolean = false;
  _schemaManager: schema.SchemaManager;
  _queryManager: ?query.QueryManager;

  _config: GQLConfig;

  _onError: (err: Error) => void;

  constructor(options: ?Options) {
    const {
      onInit = _noop,
      onChange = _noop,
      onError = _noop,
      watchman,
      watch,
      debug: enableDebug,
      ...configOptions
    } = options || {};
    if (enableDebug) {
      (debug: any).enable();
    }

    this._onError = onError;

    try {
      this._config = new GQLConfig(configOptions);
      this._watcher = new GQLWatcher({ watchman, watch });
    } catch (e) {
      process.nextTick(() => onError(e));
      return;
    }

    this._schemaManager = new schema.SchemaManager({
      config: this._config,
      watcher: this._watcher,
    });

    this._schemaManager.onChange(onChange);
    this._schemaManager.onError(onError);
    this._schemaManager.onInit(() => {
      if (!this._config.getQuery()) {
        this._isInitialized = true;
        onInit();
        return;
      }

      const queryManager = new query.QueryManager({
        config: this._config,
        getSchema: () => this._schemaManager.getSchema(),
        watcher: this._watcher,
      });

      this._queryManager = queryManager;

      queryManager.onInit(() => {
        this._isInitialized = true;
        onInit();
      });
      queryManager.onChange(onChange);
      queryManager.onError(onError);
    });
  }

  // list all file extensions defined in gqlconfig
  getFileExtensions(): Array<string> {
    return this._config.getFileExtensions();
  }

  getSchema() {
    return this._schemaManager.getSchema();
  }

  status(): Array<GQLError> {
    if (!this._isInitialized) {
      return [];
    }
    const schemaErrors = this._schemaManager.getSchemaErrors();
    const queryErrors = this._queryManager ? this._queryManager.getErrors() : [];
    return schemaErrors.concat(queryErrors.filter((err) => Boolean(err)));
  }

  autocomplete(params: CommandParams): Array<GQLHint> {
    try {
      debug.log('\n[autocomplete]');
      debug.time('time');
      const { sourceText, sourcePath, position } = params;
      if (!this._isInitialized) {
        return [];
      }

      // codemirror instance
      let results = [];
      debug.time('match');
      const match = this._config.match(sourcePath);
      debug.timeEnd('match');
      debug.log('FileType:', match && match.type);

      debug.time('autocomplete');
      if (match && match.type === 'schema') {
        results = schema.commands.getHintsAtPosition(
          this._schemaManager.getSchema(),
          sourceText,
          position,
        );
      }

      if (match && match.type === 'query') {
        results = query.commands.getHintsAtPosition(
          this._schemaManager.getSchema(),
          sourceText,
          position,
          match.opts,
        );
      }
      debug.timeEnd('autocomplete');
      debug.timeEnd('time');
      return results;
    } catch (err) {
      this._onError(err);
      return [];
    }
  }

  getDef(params: CommandParams): ?DefLocation {
    debug.log('\n[getDef]');
    debug.time('time');
    const { sourceText, sourcePath, position } = params;
    if (!this._isInitialized) {
      return undefined;
    }
    try {
      let defLocation = null;
      debug.time('match');
      const match = this._config.match(sourcePath);
      debug.timeEnd('match');
      debug.log('FileType:', match && match.type);

      debug.time('getDef');
      if (match && match.type === 'schema') {
        defLocation = schema.commands.getDefinitionAtPosition(
          this._schemaManager.getSchema(),
          sourceText,
          position,
        );
      }

      if (match && match.type === 'query') {
        defLocation = query.commands.getDefinitionAtPosition(
          this._schemaManager.getSchema(),
          sourceText,
          position,
          match.opts.parser,
        );
      }
      debug.timeEnd('getDef');
      debug.timeEnd('time');
      return defLocation;
    } catch (err) {
      this._onError(err);
      return undefined;
    }
  }

  findRefs(params: CommandParams): Array<DefLocation> {
    try {
      debug.log('\n[findRefs]');
      debug.time('time');
      const { sourceText, sourcePath, position } = params;
      if (!this._isInitialized) {
        return [];
      }

      let refLocations = [];
      debug.time('match');
      const match = this._config.match(sourcePath);
      debug.timeEnd('match');
      debug.log('FileType:', match && match.type);

      debug.time('findRefs');
      if (match && match.type === 'schema') {
        refLocations = schema.commands.findRefsOfTokenAtPosition(
          this._schemaManager.getSchema(),
          sourceText,
          position,
        );
      }
      debug.timeEnd('findRefs');

      debug.timeEnd('time');
      // @TODO query not implemented
      return refLocations;
    } catch (err) {
      this._onError(err);
      return [];
    }
  }

  getInfo(params: CommandParams): ?GQLInfo {
    try {
      debug.log('\n[getInfo]');
      debug.time('time');
      const { sourcePath, sourceText, position } = params;
      if (!this._isInitialized) {
        return undefined;
      }

      let info = null;
      debug.time('match');
      const match = this._config.match(sourcePath);
      debug.timeEnd('match');
      debug.log('FileType:', match && match.type);

      debug.time('getInfoAtToken');
      if (match && match.type === 'schema') {
        info = schema.commands.getInfoOfTokenAtPosition(
          this._schemaManager.getSchema(),
          sourceText,
          position,
        );
      }

      if (match && match.type === 'query') {
        info = query.commands.getInfoOfTokenAtPosition(
          this._schemaManager.getSchema(),
          sourceText,
          position,
          match.opts,
        );
      }
      debug.timeEnd('getInfoAtToken');

      debug.timeEnd('time');
      return info;
    } catch (err) {
      this._onError(err);
      return undefined;
    }
  }
}
