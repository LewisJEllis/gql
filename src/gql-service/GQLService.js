/* @flow */
import GQLSchemaService from 'gql-schema-service';
import GQLQueryService from 'gql-query-service';
import GQLConfig from 'gql-config';
import GQLWatcher from 'gql-watcher';
import GQLBaseService from 'gql-shared/GQLBaseService';

import debug from 'gql-shared/debug';

import {
  type GQLHint,
  type GQLInfo,
  type GQLLocation,
  type GQLPosition,
} from 'gql-shared/types';

import { type GQLError } from 'gql-shared/GQLError';

type Options = $Exact<{
  cwd?: string,

  // watch
  watchman?: boolean,
  watch?: boolean,

  // debug
  debug?: boolean,
}>;

type CommandParams = $Exact<{
  sourcePath: string,
  sourceText: string,
  position: GQLPosition,
}>;

export type { Options as GQLServiceOptions };

export default class GQLService extends GQLBaseService {
  _schemaService: GQLSchemaService;
  _queryService: ?GQLQueryService;

  _config: GQLConfig;
  _watcher: GQLWatcher;

  constructor(_options: ?Options) {
    super();
    const options = _options || {};

    // enable debug mode
    if (options.debug) {
      debug.enable();
    }

    this._config = new GQLConfig({ cwd: options.cwd || process.cwd() });
    this._watcher = new GQLWatcher({
      watchman: options.watchman,
      watch: options.watch,
    });

    // setup schema service
    this._schemaService = new GQLSchemaService({
      config: this._config,
      watcher: this._watcher,
    });
    this._schemaService.onChange(this._triggerChange);
    this._schemaService.onError(this._triggerError);

    // setup query service
    if (this._config.getQueryConfig()) {
      const queryService = new GQLQueryService({
        config: this._config,
        getSchema: () => this._schemaService.getSchema(),
        watcher: this._watcher,
      });
      queryService.onChange(this._triggerChange);
      queryService.onError(this._triggerError);
      this._queryService = queryService;
    }
  }

  async _handleStart() {
    // start schema service
    await this._schemaService.start();
    if (this._queryService) {
      await this._queryService.start();
    }
  }

  async _handleStop() {
    await this._schemaService.stop();
    if (this._queryService) {
      await this._queryService.stop();
    }
  }

  getSchema(): GQLSchema {
    return this._schemaService.getSchema();
  }

  getConfig(): GQLConfig {
    return this._config;
  }

  status(): Array<GQLError> {
    if (!this._isRunning) {
      return [];
    }
    try {
      const schemaErrors = this._schemaService.getSchemaErrors();
      const queryErrors = this._queryService
        ? this._queryService.getErrors()
        : [];
      return schemaErrors.concat(queryErrors.filter(err => Boolean(err)));
    } catch (err) {
      this._triggerError(err);
      return [];
    }
  }

  autocomplete(params: CommandParams): Array<GQLHint> {
    return this._catchThrownErrors(() => {
      debug.log('\n[autocomplete]');
      debug.time('time');
      if (!this._isRunning) {
        return [];
      }

      // codemirror instance
      let results = [];
      debug.time('fileConfig');
      const fileConfig = this._config.getFileConfig(params.sourcePath);
      debug.timeEnd('fileConfig');
      debug.log('FileType:', fileConfig && fileConfig.type);

      debug.time('autocomplete');
      if (fileConfig && fileConfig.type === 'schema') {
        results = this._schemaService.getHintsAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      if (this._queryService && fileConfig && fileConfig.type === 'query') {
        results = this._queryService.getHintsAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      debug.timeEnd('autocomplete');
      debug.timeEnd('time');
      return results;
    }, []);
  }

  getDef(params: CommandParams): ?GQLLocation {
    return this._catchThrownErrors(() => {
      debug.log('\n[getDef]');
      debug.time('time');
      if (!this._isRunning) {
        return null;
      }

      let defLocation = null;
      debug.time('fileConfig');
      const fileConfig = this._config.getFileConfig(params.sourcePath);
      debug.timeEnd('fileConfig');
      debug.log('FileType:', fileConfig && fileConfig.type);

      debug.time('getDef');
      if (fileConfig && fileConfig.type === 'schema') {
        defLocation = this._schemaService.getDefinitionAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }

      if (this._queryService && fileConfig && fileConfig.type === 'query') {
        defLocation = this._queryService.getDefinitionAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      debug.timeEnd('getDef');
      debug.timeEnd('time');
      return defLocation;
    }, null);
  }

  findRefs(params: CommandParams): Array<GQLLocation> {
    return this._catchThrownErrors(() => {
      debug.log('\n[findRefs]');
      debug.time('time');
      if (!this._isRunning) {
        return [];
      }

      let refLocations = [];
      debug.time('fileConfig');
      const fileConfig = this._config.getFileConfig(params.sourcePath);
      debug.timeEnd('fileConfig');
      debug.log('FileType:', fileConfig && fileConfig.type);

      debug.time('findRefs');
      if (fileConfig && fileConfig.type === 'schema') {
        refLocations = this._schemaService.findRefsOfTokenAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      debug.timeEnd('findRefs');

      debug.timeEnd('time');
      // @TODO query not implemented
      return refLocations;
    }, []);
  }

  getInfo(params: CommandParams): ?GQLInfo {
    return this._catchThrownErrors(() => {
      debug.log('\n[getInfo]');
      debug.time('time');
      if (!this._isRunning) {
        return null;
      }

      let info = null;
      debug.time('fileConfig');
      const fileConfig = this._config.getFileConfig(params.sourcePath);
      debug.timeEnd('fileConfig');
      debug.log('FileType:', fileConfig && fileConfig.type);

      debug.time('getInfoAtToken');
      if (fileConfig && fileConfig.type === 'schema') {
        info = this._schemaService.getInfoOfTokenAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }

      if (this._queryService && fileConfig && fileConfig.type === 'query') {
        info = this._queryService.getInfoOfTokenAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      debug.timeEnd('getInfoAtToken');

      debug.timeEnd('time');
      return info;
    }, null);
  }
}
