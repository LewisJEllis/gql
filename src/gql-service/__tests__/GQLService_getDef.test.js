/* @flow */
import code from 'gql-test-utils/code';
import path from 'path';
import { createTempFiles } from 'gql-test-utils/file';
import GQLService from '../GQLService';

describe('Schema: getDef', () => {
  it('works in schema files', async () => {
    const dir = createTempFiles({
      '.gqlconfig': `
        {
          schema: {
            files: 'schema/*.gql',
          }
        }
      `,
      'schema/schema.gql': `
        type Query {
          viewer: Viewer
        }

        type Viewer {
          name: string
        }
      `,
    });
    const gql = new GQLService({
      cwd: dir,
      watch: false,
    });

    gql.onError(err => {
      throw err;
    });

    await gql.start();

    expect(
      gql.getDef({
        sourcePath: path.join(dir, 'schema/user.gql'),
        ...code(`
          type User {
            viewer: Viewer
            #---------^
          }
        `),
      }),
    ).toMatchSnapshot();
  });

  it('works in schema files', () => {
    const dir = createTempFiles({
      '.gqlconfig': `
        {
          schema: {
            files: 'schema/*.gql',
          }
        }
      `,
      'schema/schema.gql': `
        type Query {
          viewer: Viewer
        }

        type Viewer {
          name: string
        }
      `,
    });
    const gql = new GQLService({
      cwd: dir,
      watch: false,
    });

    gql.onError(err => {
      throw err;
    });

    const run = () =>
      gql.getDef({
        sourcePath: path.join(dir, 'schema/user.gql'),
        ...code(`
          type User {
            viewer: Viewer
            #---------^
          }
        `),
      });

    expect(run).not.toThrow();
    expect(run()).toBe(null);
  });
});

describe('Query: getDef', () => {
  it('works in query files', async () => {
    const dir = createTempFiles({
      '.gqlconfig': `
        {
          schema: {
            files: 'schema/*.gql',
          },
          query: {
            files: [
              {
                match: 'query/*.gql',
              },
            ]
          }
        }
      `,
      'schema/schema.gql': `
        type Query {
          viewer: Viewer
        }

        type Viewer {
          name: string
        }
      `,
    });
    const gql = new GQLService({
      cwd: dir,
      watch: false,
    });

    gql.onError(err => {
      throw err;
    });
    await gql.start();

    expect(
      gql.getDef({
        sourcePath: path.join(dir, 'query/user.gql'),
        ...code(`
          fragment test on Viewer {
              #-------------^
            name
          }
        `),
      }),
    ).toMatchSnapshot();
  });
});
