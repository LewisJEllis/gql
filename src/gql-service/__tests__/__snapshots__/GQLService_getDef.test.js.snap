// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query: getDef works in query files 1`] = `
Object {
  "end": Object {
    "column": 10,
    "line": 8,
  },
  "path": "/tmp/gql-test/bb63494cd671b99c58779373ad9ef96a71dc250e/schema/schema.gql",
  "start": Object {
    "column": 9,
    "line": 6,
  },
}
`;

exports[`Schema: getDef works in schema files 1`] = `
Object {
  "end": Object {
    "column": 10,
    "line": 8,
  },
  "path": "/tmp/gql-test/4dae3772d2ca36e740891e5aa5ddb39c57c0712b/schema/schema.gql",
  "start": Object {
    "column": 9,
    "line": 6,
  },
}
`;
